bluetooth: fixed the issue where the thread_loop_work_sync semaphore …

liuX10 opened the pull request:
…blocking when client_loop without run.

bug: v/78154

If bluetoothd initialization fails, it causes the client_loop to execute the deinit operation before it runs, which blocks thread_loop_work_sync and consequently blocks the bluetooth_create_instance API.

gzh-terry commented on service/ipc/socket/src/bt_socket_client.c:
Any protections on `do_in_thread_loop_sync` as well?

chejinxian reviewed (approved):
If the client_loop thread is not running, the corresponding worker threads will also be unable to run.
```
    do {
        ins->peer_fd = bt_socket_client_connect(family, name, cpu, port);
        if (ins->peer_fd <= 0 && !retry) {
            /* connect fail, go out */
            BT_DFX_IPC_CONN_ERROR(BT_DFXE_CLIENT_CONNECT_FAIL, BT_DFXE_FILE_DESCRIPTOR_ERROR);
            bt_socket_client_deinit(ins);
            return BT_STATUS_PARM_INVALID;
        } else if (ins->peer_fd <= 0) {
            /* connect fail, retry after sleep 100ms */
            usleep(CLIENT_DELAY_MS(retry) * 1000);
            continue;
        } else {
            /* success, goto next step */
            break;
        }
    } while (retry--);

    poll = thread_loop_poll_fd(ins->client_loop, ins->peer_fd, UV_READABLE | UV_DISCONNECT,
        bt_socket_client_handle_event, ins);
    if (poll == NULL) {
        bt_socket_client_deinit(ins);
        return BT_STATUS_PARM_INVALID;
    }

    ins->poll = poll;

    thread_loop_run(ins->client_loop, true, "bt_client");


```

chengkai15 commented on service/ipc/socket/src/bt_socket_client.c:
```diff
@@ -516,7 +516,8 @@ void bt_socket_client_deinit(bt_instance_t* ins)
     /* Dispatch an empty work to ensure all pending work have completed, while
     `bt_socket_sync_close` guarantees no new work will enter the queue. */
-    thread_loop_work_sync(ins->client_loop, NULL, NULL, NULL);
+    if (uv_loop_alive(ins->client_loop))
```
why do_in_thread_loop_sync would not fail in that case？

liuX10 replied:
because now ins->poll == NULL, another path will not be taken.
```
if (!ins->poll)
        bt_socket_sync_close(ins);
```

huangyulong3 commented on service/ipc/socket/src/bt_socket_client.c:
```diff
@@ -516,7 +516,8 @@ void bt_socket_client_deinit(bt_instance_t* ins)
 
     /* Dispatch an empty work to ensure all pending work have completed, while
```
Please change the comment, explain why add the juge here.

huangyulong3 reviewed (approved):
LGTM.

liuX10 committed:
bluetooth: fixed the issue where the thread_loop_work_sync semaphore blocking when client_loop without run.

bug: v/78154

If bluetoothd initialization fails, it causes the client_loop to execute the deinit operation before it runs, which blocks thread_loop_work_sync and consequently blocks the bluetooth_create_instance API.

Signed-off-by: liuxiang18 <liuxiang18@xiaomi.com>
